### PR TITLE
FEATURE: Allow setting authentication provider on new user creation in user backend module

### PR DESCRIPTION
**Pull request by dlubitz**
Allow setting authentication provider on new user creation in user backend module.

If less then two providers are given, the selector does not appear and the default authentication provider is used. Same as no authentication provider is explicit selected.

### CODE REVIEW

**Comment by dlubitz** [comment]
screen

Unattractive is at the moment the TYPO3SetupProvider. Maybe it make sence to filter providers by provider type `PersistedUsernamePasswordProvider`?

File: TYPO3.Neos/Classes/TYPO3/Neos/Controller/Module/Administration/UsersController.php [hunk 1]
```diff
@@ -295,4 +303,18 @@ protected function assignElectronicAddressOptions()
             'electronicAddressUsageTypes' => $electronicAddressUsageTypes
         ));
     }
+
+    /**
+     * TODO: With Flow 4.0 replace getting providers with access over AuthenticationManagerInterface
```
Comment (gerhard-boden) [inline TYPO3.Neos/Classes/TYPO3/Neos/Controller/Module/Administration/UsersController.php]: Wouldn't this change be released at the same time as: https://github.com/neos/flow-development-collection/pull/404 
I suppose both changes would be merged into the next mayor release, so you can already use your getter here.

**Comment by gerhard-boden** [comment]
Thanks @dlubitz, nice addition!

Checked out your PR and created a new user. I selected `Flowpack.Neos.FrontendLogin:Frontend` but the account is still persisted as `Typo3BackendProvider`. 

Is there a technical reason why this change only covers changing the authentication provider for newly created users and not for existing? Probably not an everyday use-case but I can imagine that the provider could change, or that someone selected the wrong provider for a newly created user.

**Comment by dlubitz** [comment]
@gerhard-boden Thanks for your feedback.

I will test it again, tomorrow.

> Is there a technical reason why this change only covers changing the authentication provider for newly created users and not for existing? 
I don't know, was more a conceptional thought. It usally doesn't make sence to change the authentication provider. The account is bound to the provider, username is unique by provider etc.

Regarding the point of release ... I think this feature will release with next "normal" release (Neos 2.3 LTS), and not with Neos 3.0. Don't know if it makes sence, but I could already provide a second PR for Neos 3.0. So we don't have to "remember" to do this change?

**Comment by kitsunet** [comment]
Technically you could change the provider of course but it makes no sense. The account table should contain provider specific credentials and it might be that only this provider understands them (Depending on the provider configuraiton). Eg. if I have a UserNamePassword and a Oauth provider you cannot simple change the provider name. So IMHO it's fine if that cannot be edited.

**Comment by dlubitz** [comment]
@gerhard-boden: You were right. The selectbox had the wrong values. Is fixed now.

@kitsunet: What do you think about the second PR for Neos 3.0? Does this make sence?

**Comment by kitsunet** [comment]
As we do upmerges no second change should be needed. When this is merged early enough for 2.3 it will go in there, otherwise it would go into 3.0. If it goes into 2.3 it will be upmerged to 3.0

**Comment by dlubitz** [comment]
Sure, but for 3.0 we could already change the getting of the providers over the AuthenticationManager (which needs the https://github.com/neos/flow-development-collection/pull/404 for Flow 4.0) as we will need it then.

**Comment by gerhard-boden** [comment]
@dlubitz thanks for the update, will test again in the evening or latest by tomorrow!

Here's what i don't get: Why would neos/flow-development-collection#404 not be merged into Flow 3.3? Since it's not really a delicate change as far as i can tell.

**Comment by dlubitz** [comment]
Because of the breaking change in the interface. These changes can just be released with a major release.

**Comment by gerhard-boden** [comment]
Oh okay, obviously i did not understand the extend of that change in particular. I thought you only added a setter to the interface and one class that implements that interface.

Of course everyone who extended that interface in their custom code would have to add that getter, right.

**Comment by kitsunet** [comment]
You cannot expect that, if there is an interface it cannot change in a minor version.

**Comment by gerhard-boden** [comment]
Retested today, can confirm that it works :+1: 

Thanks for answering all my questions guys!

**Comment by dfeyer** [comment]
> Technically you could change the provider of course but it makes no sense. The account table should contain provider specific credentials and it might be that only this provider understands them (Depending on the provider configuraiton).

Must be done differently as @kitsunet say earlier, an authprovider can require different settings than a username password. So i prefer something like a wizard:

1. Edit a user
2. Click add an account
3. The wizard ask to select a Provider (provider list can be filtered by Role, to limit some Provider to Manager, Admin, ...)
4. The wizard select the proper Form template to fullfil the requirement for the selected provider

**Comment by dlubitz** [comment]
Sure, that's the pretty deluxe version. But for meanwhile I think this is a better solution than having none of them. For now, I would keep it simple and also filter the FileBasedSimpleKeyProvider and only support PersistedUsernamePasswordProvider.

**Comment by kitsunet** [comment]
If this could be rebased we could get this in. @dlubitz maybe? :)

**Comment by dlubitz** [comment]
Sure @kitsunet. What is the target branch?

**Comment by kitsunet** [comment]
Master is just fine, it's a feature after all. :) Thanks a lot.

**Comment by dlubitz** [comment]
@kitsunet Done.

**Comment by gerhard-boden** [comment]
Would merge this today if no one objects. 
Thanks again @dlubitz for holding on this long

**Pull request merged**